UBL for DELTA

oldmcg opened the pull request:
Submit for review, enable UBL for DELTA.
Hopefully someone can tell me how how to squash commits without just making more commits.

oldmcg committed:
UBL for DELTA

oldmcg committed:
UBL for DELTA

oldmcg committed:
UBL for DELTA

oldmcg committed:
UBL for DELTA

oldmcg committed:
fix ubl without fade

oldmcg committed:
Merge branch '1.1.x' of https://github.com/oldmcg/Marlin into 1.1.x

oldmcg committed:
Merge branch '1.1.x' of https://github.com/oldmcg/Marlin into 1.1.x

oldmcg committed:
Merge branch '1.1.x' of https://github.com/oldmcg/Marlin into 1.1.x

oldmcg committed:
typo

Bob-the-Kuhn commented:
There's a section in the [Marlin wiki](http://marlinfw.org/docs/development/getting_started_pull_requests.html) that goes over squashing commits.

oldmcg committed:
another typo

Roxy-3D commented:
> Hopefully someone can tell me how how to squash commits without just making more commits.

@oldmcg  One of the 'Merge' options is 'Squash and Merge'.     We will do that.  Just keep committing until everybody feels good.  And we will turn it into one massive change when we hit the button.

Let's just review it for things that might affect non-Delta printers...  Because right now... Delta's don't work at all!

Seriously guys...   I've been working with him for a while on this and he knows his stuff!  There is a good chance that when this gets merged, Deltas can 'Limp along' !!!!

@oldmcg    THANK YOU SO MUCH!!!!!!

FHeilmann commented on Marlin/Conditionals_post.h:
```diff
@@ -780,11 +780,11 @@
+    #define MIN_PROBE_X (-10000)
+    #define MAX_PROBE_X ( 10000)
+    #define MIN_PROBE_Y (-10000)
+    #define MAX_PROBE_Y ( 10000)
```
Does this solve any specific problem? Otherwise it seems fishy to me.

oldmcg replied:
The problem here is that MIN_PROBE_X and friends are used with preprocessor math, meaning they cannot be float.  But on delta, DELTA_PRINTABLE_RADIUS is a float (at least in all the sample delta config files), and MIN_PROBE_X etc are defined relative to DELTA_PRINTABLE_RADIUS.

So two options -- either require DELTA_PRINTABLE_RADIUS to be integer which might break existing user config.h or ignore all these preprocessor checks for delta and require the user to specify manual MIN_PROBE_X etc that are integers without being able to preprocessor-verify that they are within float DELTA_PRINTABLE_RADIUS, which is kind of silly anyway because the corners are still unreachable, but that is handled now within the code.

For deltas, the true probeable area is actually a circle (printable_radius) minus a crescent (offset of probe from nozzle).  This is reflected in the new position_is_reachable_by_probe for deltas.  So delta won't probe off the bed (at least for UBL) even if the MIN_PROBE_X etc are outside the radius.  

Personally I think MIN_PROBE_X etc should simply be undefined for DELTA completely and leave that definition/verification mechanism for cartesian-only.  Delta probe "grid" for UBL can be defined as the square that encompasses the printable radius.  For other bed leveling systems on delta (I think BILINEAR is the only one) I'm not certain but I think the probabeable rectangle must fit completely within the printable and probeable region of the bed.  I could change that, but it would fall outside UBL_DELTA semantics.

I'm open to suggestions.

oldmcg replied:
I changed the delta MIN_PROBE_X etc defines to be PRINTABLE_RADIUS based again, and removed the sanity checks for range-checking the UBL_PROBE_PT values for DELTA.  Hopefully this better reflects reality.

oldmcg replied:
Oops, scara is square bed, working on fix for that.

oldmcg replied:
SCARA with UBL still has problems unrelated to my changes.  UBL requires NEWPANEL which the SCARA default configuration.h does not provide (LCD errors when building with UBL).  

So for now UBL+SCARA disabled in sanitycheck.h but all the functionality for SCARA should be in UBL_DELTA, but unable to build to verify.

oldmcg committed:
Better fix for compile time check of UBL_PROBE_PT_xxx on DELTA

oldmcg committed:
Fixes for SCARA but SCARA still won't build with UBL for other reasons, so disabled in sanitycheck.h

Roxy-3D commented:
In the details there are a bunch of M100 warnings.   I believe those have been fixed in the main branch.   It probably is best to leave those alone just because any changes to fix them will cause conflicts with this Pull Request.  (At least...  That's what I think will happen.)

oldmcg commented:
I didn't touch the M100 file, any changes there should be independent of UBL_DELTA.

Roxy-3D commented:
Right!   I was concerned that you are looking at the 'Travis Details' and will keep seeing M100 warnings and get annoyed with them and fix them.   I've been bitten by doing things like that.   If you make a change where a change has already happened...   You end up with conflicts.   And some times those conflicts can't be resolved.

bgort commented:
I can confirm that all of the M100 compiler warnings are fixed.  @thinkyhead got most a few days back, and I got the rest yesterday (I think it was yesterday).

oldmcg commented:
Should I merge to current in my branch or leave it alone?

Roxy-3D commented:
I would leave it alone...   I know this sounds paranoid....  But when I do that, I pick up conflicts I can't resolve.    Right now...  We have your changes captured.   Let's not push our luck!!!

But it would be safe to download a .ZIP file of the current if you want to check things out.  You just don't want to make a change to an area that has already had a change.  That causes huge head aches.

Do you have everything uploaded and the way you want it?  If so...  I'll start scanning through things and make sure anything without #ifdef's isn't going to cause hiccups.    And if everybody can get comfortable with that conclusion, we want to merge as soon as possible because right now other changes can cause conflicts and sink the ship.

Roxy-3D commented:
@oldmcg    I think this breaks Scara's ???    We could put the #if DELTA stuff in front...  And then have an #elif IS_KINEMATIC with everything just the way it was.
```cpp
-  #if IS_KINEMATIC
 -    // Check for this in the code instead
 -    #define MIN_PROBE_X X_MIN_POS
 -    #define MAX_PROBE_X X_MAX_POS
 -    #define MIN_PROBE_Y Y_MIN_POS
 -    #define MAX_PROBE_Y Y_MAX_POS
 +  #if ENABLED(DELTA)
 +    // These will be further constrained in code, but UBL_PROBE_PT values
 +    // cannot be compile-time verified within the radius.
 +    #define MIN_PROBE_X (-DELTA_PRINTABLE_RADIUS)
 +    #define MAX_PROBE_X ( DELTA_PRINTABLE_RADIUS)
 +    #define MIN_PROBE_Y (-DELTA_PRINTABLE_RADIUS)
 +    #define MAX_PROBE_Y ( DELTA_PRINTABLE_RADIUS)
    #else
```
or actually... what would be cleaner would be 
```cpp
#if IS_KINEMATIC
   #if ENABLED(DELTA)
     // These will be further constrained in code, but UBL_PROBE_PT values
     // cannot be compile-time verified within the radius.
     #define MIN_PROBE_X (-DELTA_PRINTABLE_RADIUS)
     #define MAX_PROBE_X ( DELTA_PRINTABLE_RADIUS)
     #define MIN_PROBE_Y (-DELTA_PRINTABLE_RADIUS)
     #define MAX_PROBE_Y ( DELTA_PRINTABLE_RADIUS)
  #else  // other KINEMATIC's
    #define MIN_PROBE_X X_MIN_POS
    #define MAX_PROBE_X X_MAX_POS
    #define MIN_PROBE_Y Y_MIN_POS
    #define MAX_PROBE_Y Y_MAX_POS
  #endif
#else  // not KINEMATIC

oldmcg commented:
That's everything I have.  Had to disable scara because UBL requires NEWPANEL LCD and scara example config doesn't have one.  But the scara stuff is in there, untested.  

The biggest outside-of-UBL change is the position_is_reachable stuff, moved from marlin_main to inline in marlin.h so could be accessed from ubl and elsewhere.  General use of WITHIN and constrain to limit motion should use position_is_reachable instead -- I changed the places that were obvious.

Also, planner leveling is enabled with UBL_DELTA so things like G2/G3 arc and set_position should work as expected with UBL, and I added unapply_leveling to be keep things functionally parallel with other leveling systems.

Cartesian can optionally use UBL_DELTA which will switch from the existing cartesian-only ubl_line_to to the new ubl segmented routine result in smaller segments (1mm default) following the interpolated mesh between cell boundaries.

Fixed a couple of related minor bugs, like scara calling final segment on "logical" instead of "ltarget" and call to prepare_move_to_destination_dualx() falling through to prepare_move_to_destination_cartesian if it returned false due to faulty #ifdef logic.

Oh, and delta will now quietly ignore do_blocking_move_to targets outside the printable radius.  I hated seeing my delta bind up against physical limits, bending rods and jumping belt teeth.  Kind of like SOFTWARE_ENDSTOPS.

Roxy-3D commented:
Do agree with the suggestions here: https://github.com/MarlinFirmware/Marlin/pull/6686#issuecomment-300851953

oldmcg commented:
Yes, see the last commit with related comments under FHeilmann review thread.

Roxy-3D commented:
Thank You!   Oh!   I didn't realize I had old stuff!    That is one of the problems with not Squashing the Pull Request.   But I didn't want you to have to jump through extra hoops.

Roxy-3D commented:
@oldmcg   Hey!   One quick question...  How do you envision that UBL_DELTA gets defined?    It might make sense to add a quick check in conditionals_post.h that says:
```cpp
#if  ENABLED(AUTO_BED_LEVELING_UBL) && ENABLED(DELTA)
  #define UBL_DELTA
#endif
```
I've looked over everything...  And I think I feel comfortable merging.   I also built and loaded my Cartesian printer with the code and it is running UBL to do a print.  Everything looks good there!

So...  If you want to sneak that automatic defining of UBL_DELTA into the mix...   I'll hold off on the merge.

Roxy-3D commented:
We have UBL for Delta's !!!!!!

PS.  @oldmcg    I'm adding the automatic definition of UBL_DELTA right now to make things easier for the Delta users.

bgort commented:
FYI:  this was merged into the '1.1.x' branch, not bugfix .. not sure if that was intentional, but I just did a fetch and saw the changes to that branch and thought I'd mention, just in case.

oldmcg commented:
Yes, since UBL on DELTA didn't compile or work before, I see no reason not to automatically enable UBL_DELTA.  You beat me to it -- I added your suggestion after you merged.

Note that UBL_DELTA can optionally be specified for CARTESIAN too -- that is the scenario I'm curious about because I don't have a cartesian machine to test it.  If feedrate is too fast it might stutter -- delta_segments_per_segment is not defined for cartesians so I just used fixed 1mm segment length as default for cartesian.

oldmcg commented:
Yes, Roxy's merge happened before I pushed the equivalent change to the PR.

Roxy-3D commented:
> FYI: this was merged into the '1.1.x' branch, not bugfix .. not sure if that was intentional, but I just did a fetch and saw the changes to that branch and thought I'd mention, just in case.

Yikes!!!!   How do we fix that?    I don't think we want to revert it because that will cause the PR to be lost.

Roxy-3D commented:
@oldmcg   Can you make the equivilent PR for bugfix_v1.1.0   And then I can revert the other one?

How do we recover?

bgort commented:
I think we can pull down a patch at:
https://github.com/MarlinFirmware/Marlin/pull/6686.patch

That'll allow us to save the changes so they can't be lost.  Then we can revert the 1.1.x commit, checkout a new branch against bugfix, then `cat <patchfile> | git apply` .. I think.

bgort commented:
Let me try the new branch .. one sec.

Roxy-3D commented:
@bgort    Can you do this?   My Git skills are so poor I'll probably screw things up worse.

oldmcg commented:
Oops, did I initially fork from wrong branch?  I thought I forked from bugfix-1.1.x

bgort commented:
My guess is when you created the PR you specified the wrong branch.  It defaults to 1.1.x.

oldmcg commented:
So even though I forked from bugfix-1.1.x when I created PR it switched to default 1.1.x?

bgort commented:
Yes.  It uses the default branch for the repo.

oldmcg commented:
Should I just create a new PR to bugfix-1.1.x from my current fork?  Or is there some implicit linkage between my fork and 1.1.x now?

Roxy-3D commented:
> So even though I forked from bugfix-1.1.x when I created PR it switched to default 1.1.x?

I could be wrong on this...  But when you fork...  You get all the branches.  It doesn't matter where you are pointed at the moment you fork.   And once you fork, you pick up everything including how the default branch is set.

When I fork, the first thing I do is go change the default branch and the name of the fork.

bgort commented:
This is giving me grief.  @oldmcg:  I assume you still have your modified branch on github?   If so, just create a new PR against bugfix-1.1.x.

Roxy-3D commented:
Let's see if bgort can fix it.   But if he can't...  Yeah, we will need to recreate the Pull Request.

bgort commented:
I was able to apply the downloaded patch to a new branch, but it didn't like a few things, so I'm wondering if there are conflicts with bugfix that weren't obvious when the PR was against 1.1.x.

Roxy-3D commented:
Unknown!    What is it complaining about?

bgort commented:
```
patching file Marlin/Marlin.h
Hunk #1 FAILED at 429.
1 out of 1 hunk FAILED -- saving rejects to file Marlin/Marlin.h.rej
...
patching file Marlin/ubl.h
Hunk #2 succeeded at 122 (offset -1 lines).
Hunk #3 FAILED at 163.
Hunk #4 succeeded at 330 (offset 2 lines).
1 out of 4 hunks FAILED -- saving rejects to file Marlin/ubl.h.rej
...
patching file Marlin/ubl_G29.cpp
Hunk #1 succeeded at 439 (offset 4 lines).
Hunk #2 succeeded at 463 (offset 4 lines).
Hunk #3 succeeded at 482 (offset 4 lines).
Hunk #4 succeeded at 522 (offset 4 lines).
Hunk #5 succeeded at 736 (offset 4 lines).
Hunk #6 succeeded at 751 (offset 4 lines).
Hunk #7 succeeded at 955 (offset 10 lines).
Hunk #8 FAILED at 1021.
Hunk #9 succeeded at 1378 (offset 4 lines).
Hunk #10 succeeded at 1436 (offset 4 lines).
Hunk #11 FAILED at 1446.
Hunk #12 succeeded at 1460 (offset 3 lines).
Hunk #13 succeeded at 1517 (offset 3 lines).
2 out of 13 hunks FAILED -- saving rejects to file Marlin/ubl_G29.cpp.rej
patching file Marlin/ubl_motion.cpp
Hunk #4 FAILED at 225.
Hunk #9 succeeded at 426 (offset -2 lines).
Hunk #10 succeeded at 455 (offset -4 lines).
Hunk #11 succeeded at 472 (offset -2 lines).
1 out of 11 hunks FAILED -- saving rejects to file Marlin/ubl_motion.cpp.rej
```

bgort commented:
I can go and find the conflicts in the .rej files, but I'm pretty sure when oldmcg does a new PR it'll be obvious where the problems are.

Roxy-3D commented:
OK...   Let's see what happens when oldmcg does a new PR against bugfix_v1.1.0    I guess we should leave the current merge until we hear what happens.  Because if I revert and he can't make the PR, we will lose everything.

I tell you....  Git is like a mine field.   You think you are doing everything right, and *BLAM*, something horrible happens.

bgort commented:
I have the changes saved, just in case.

oldmcg commented:
I am creating new PR now against bugfix-1.1.x.  It might have build problems initially -- I will look at the travis log and fix them.

oldmcg commented:
No, it looks like my branch is against 1.1.x, not sure how that happened.

I'm not worried about losing the dozen files I modified -- I have them archived.

If I rebase my local branch to bugfix-1.1.x will that attempt to reapply all my changes on top of it?

Roxy-3D commented:
> If I rebase my local branch to bugfix-1.1.x will that attempt to reapply all my changes on top of it?

In theory, I think that works...  But if you have any changes where other changes have happened...  You get conflicts.   And I have gotten into situations where they can't be resolved.

We may as well try it...   But be sure your directory is safe.  Maybe that you have multiple copies of it.  I'm worried about losing the changes!

bgort commented:
It should, yes, but that could be messy as there have been a lot of commits between 1.1.x and bugfix.  

:crossed_fingers:

bgort commented:
As long as this PR isn't deleted, we'll always have access to the .patch, and I have a copy locally.  Nothing to worry about.

Roxy-3D commented:
OK...  If the rebase works...   We set that aside.    And then do a new fork of the current bugfix_v1.1.0.   And then drop his rebased files on top of that.   Git should just pick up the changes to make a new, clean PR, right?

bgort commented:
If the rebase works he can either create a new local branch, push that to github, then create a new PR; or, he can just push -f (force) overwrite his existing branch on github, then create a new PR from there.

oldmcg commented:
It's going to take me a little while -- don't hold your breath.  I've got a few merge conflicts but should be fairly straightforward to work them out.  Oddly I also got merge conflicts in files I haven't touched.  I don't know the git syntax to say "just take the #head version of this file from the new branch".  git checkout suchfile results in error: 'suchfile' is unmerged.  I'll figure it out, but might be later tonight or in the morning before I get the new PR posted.

Sorry for the hassle.  Git is unlike any other SCM I've used

Roxy-3D commented:
OK...   I'll revert!    I'm sorry I didn't notice the branch error!!!

bgort commented:
You might want to do a `git fetch upstream` prior to the rebase if you're seeing conflicts in files you didn't modify.  I assume you're rebasing off of 'upstream/bugfix-1.1.x' ?

>Sorry for the hassle. Git is unlike any other SCM I've used

It takes quite a while to learn git - at least it did for me.

Roxy-3D commented:
@bgort   I can't find the 'Revert' button on that commit?   Am I missing something?

Maybe you better do it!     I'm losing my confidence!

bgort commented:
I've never reverted on github, but I'll do it.... 

Standby...

Roxy-3D commented:
Thank You.

bgort commented:
No problem.  The button was up next to the merge notification in this PR.

Roxy-3D commented:
Well...  The good news is this set of changes works on Cartesian Printers.   My print finished and it is fine!

bgort commented:
Very cool...
